File: problems/C++/08. String to Integer (atoi).cpp
class Solution {
public:
    int myAtoi(string s) {
        int imax = INT_MAX;
        int imin = INT_MIN;
        
        int i = 0;
        int n = s.length();
        
        while (i < n && s[i] == ' ') {
            i++;
        }
        
        int sgn = 1;
        if (i < n && (s[i] == '+' || s[i] == '-')) {
            sgn = (s[i] == '-') ? -1 : 1;
            i++;
        }
        
        int res = 0;
        while (i < n && isdigit(s[i])) {
            int dig = s[i] - '0';
            
            if (res > imax / 10 || (res == imax / 10 && dig > 7)) {
                return (sgn == 1) ? imax : imin;
            }
            
            res = res * 10 + dig;
            i++;
        }
        
        return sgn * res;
    }
};

File: problems/Rust/08. String to Integer (atoi).rs
impl Solution {
    pub fn my_atoi(s: String) -> i32 {
        let imax = i32::MAX;
        let imin = i32::MIN;
        let b = s.as_bytes();
        let n = b.len();
        let mut i: usize = 0;

        while i < n && b[i] == b' ' {
            i += 1;
        }

        let mut sgn: i32 = 1;
        if i < n && (b[i] == b'+' || b[i] == b'-') {
            if b[i] == b'-' {
                sgn = -1;
            }
            i += 1;
        }
        
        let mut res: i32 = 0;
        while i < n && b[i] >= b'0' && b[i] <= b'9' {
            let dig = (b[i] - b'0') as i32;

            if res > imax / 10 || (res == imax / 10 && dig > 7) {
                return if sgn == 1 { imax } else { imin };
            }
            
            res = res * 10 + dig;
            i += 1;
        }

        sgn * res
    }
}